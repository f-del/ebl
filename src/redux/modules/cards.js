import update from "immutability-helper";

const CREATE = "CARD/CREATE";
const START = "CARD/START";

const card = title => ({
  Title: title,
  Status: cardStatus.Todo,
  Type: cardType.Task
});

export const cardStatus = {
  Todo: "TODO",
  Inprogress: "INPROGRESS"
};

export const cardType = {
  Task: "TASK"
};

export const createCardSuccess = (id, card) => ({
  type: CREATE,
  payload: { Id: id, ...card }
});

export const createCard = title => {
  return (dispatch, getState, { api }) => {
    const newCard = card(title);
    return api.Tasks.Post(newCard).then(res => {
      console.log("post done,  new Id =" + res.Id);
      dispatch(createCardSuccess(res.Id, newCard));
    });
  };
};

export const startCard = id => {
  if (id === undefined) throw new Error("Id arg is mandatory");
  return {
    type: START,
    payload: {
      Id: id
    }
  };
};

const initialState = {
  list: []
};

export default function(state = initialState, action) {
  console.debug(action);
  switch (action.type) {
    case CREATE:
      return { ...state, list: [...state.list, action.payload] };
    case START:
      let idx = state.list.findIndex(c => c.Id === action.payload.Id);

      var mergeObject = {};
      mergeObject[idx] = { $merge: { Status: cardStatus.Inprogress } };

      return { ...state, list: update(state.list, mergeObject) };

    default:
      return state;
  }
}

export function getAllCards(state) {
  if (state === undefined) throw new Error("State arg is mandatory");
  return state.cards !== undefined ? state.cards.list : [];
}

export function getAllCardsInProgess(state) {
  return getAllCards(state).filter(isCardStatusInProgess);
}
export function getAllCardsTodo(state) {
  return getAllCards(state).filter(isCardStatusToDo);
}

/* Functionnal helper */
function isCardStatusInProgess(card) {
  return card.Status === cardStatus.Inprogress;
}
function isCardStatusToDo(card) {
  return card.Status === cardStatus.Todo;
}
