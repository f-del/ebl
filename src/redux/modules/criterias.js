export const criteriaType = {
  BASIC: "BASIC",
  DEV: "DEV"
};

export function create(id, value) {
  return { Id: id, Value: value };
}

const initialState = {
  list: {
    BASIC: [
      {
        Id: "IdCritBas1",
        Value: false
      }
    ],
    DEV: [
      {
        Id: "IdCritDev1",
        Text: "Development",
        Value: false
      },
      {
        Id: "IdCritDev2",
        Text: "Unit testing",
        Value: false
      }
    ]
  }
};

export default function(state = initialState, action) {
  return state;
}

export function getCriteria(state, type) {
  if (type === undefined) throw new Error("Type argument is mandatory");
  return getState(state).list[type] || [];
}
function getState(state) {
  return state.criterias || {};
}
