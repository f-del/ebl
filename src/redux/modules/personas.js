/***
 *    ██████╗ ███████╗██████╗ ██╗   ██╗ ██████╗███████╗██████╗ ███████╗
 *    ██╔══██╗██╔════╝██╔══██╗██║   ██║██╔════╝██╔════╝██╔══██╗██╔════╝
 *    ██████╔╝█████╗  ██║  ██║██║   ██║██║     █████╗  ██████╔╝███████╗
 *    ██╔══██╗██╔══╝  ██║  ██║██║   ██║██║     ██╔══╝  ██╔══██╗╚════██║
 *    ██║  ██║███████╗██████╔╝╚██████╔╝╚██████╗███████╗██║  ██║███████║
 *    ╚═╝  ╚═╝╚══════╝╚═════╝  ╚═════╝  ╚═════╝╚══════╝╚═╝  ╚═╝╚══════╝
 *
 */

const initialState = {
  list: {
    e6wTJIBsAw5puwwrNEsR: {
      Name: "Feature Team",
      Details: "Meta persona, representing all collaborators working with Paul",
      Needs: [
        "To understand WHAT the Product Owner wants to be able to design collaborativly the HOW",
        "To manage easly the list of things to do to finish the most quickly the User Story"
      ],
      Picture: ""
    },
    uOCF2GLFiJl6fvxdlq4b: {
      Details: "Product Owner for a new product, work in a startup",
      Name: "Paul Deligny",
      Needs: [
        "To have a clear understanding about the product vision & objectives",
        "To write real User Stories, ie have a Customer Driven thinking",
        "To share at the right time and with the rigth level of information the User Stories with his team",
        "To have a clear vision on the job he'll have to do AND the job already or currently done"
      ],
      Picture: ""
    }
  }
};

export default function(state = initialState, action) {
  return state;
}

/***
 *    ███████╗███████╗██╗     ███████╗ ██████╗████████╗ ██████╗ ██████╗ ███████╗
 *    ██╔════╝██╔════╝██║     ██╔════╝██╔════╝╚══██╔══╝██╔═══██╗██╔══██╗██╔════╝
 *    ███████╗█████╗  ██║     █████╗  ██║        ██║   ██║   ██║██████╔╝███████╗
 *    ╚════██║██╔══╝  ██║     ██╔══╝  ██║        ██║   ██║   ██║██╔══██╗╚════██║
 *    ███████║███████╗███████╗███████╗╚██████╗   ██║   ╚██████╔╝██║  ██║███████║
 *    ╚══════╝╚══════╝╚══════╝╚══════╝ ╚═════╝   ╚═╝    ╚═════╝ ╚═╝  ╚═╝╚══════╝
 *
 */
export function getAllPersonas(state) {
  return mapToArrayPersona(getPersonasList(state)) || [];
}

function getPersonasList(state) {
  return getPersonas(validateState(state)).list || {};
}

/* Functionnal helper */
function validateState(state) {
  if (state === undefined) throw new Error("State arg is mandatory");
  return state;
}
function getPersonas(state) {
  return state.personas || {};
}

function mapToArrayPersona(list) {
  return Array.from(Object.keys(list), key => ({
    ...getPersonaId(key),
    ...list[key]
  }));
}
function getPersonaId(key) {
  return {
    Id: key
  };
}
