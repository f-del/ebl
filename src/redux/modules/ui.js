const SELECT_HYPOTHESIS = "UI/SELECT/HYPOTHESIS";
const SELECT_USERSTORY = "UI/SELECT/USER-STORY";

export const selectHypothesis = hypothesis => ({
  type: SELECT_HYPOTHESIS,
  payload: {
    Id: hypothesis.Id
  }
});

export const selectUserStory = userstory => ({
  type: SELECT_USERSTORY,
  payload: {
    Id: userstory.Id
  }
});

const initialState = {
  currentHypothesis: undefined,
  currentUserStory: undefined
};

export default function uiReducer(state = initialState, action) {
  switch (action.type) {
    case SELECT_HYPOTHESIS: {
      // currently Reducer handle Ui logic -> deselect current US on Hypothesis change
      return {
        ...state,
        currentHypothesis: action.payload.Id,
        currentUserStory: undefined
      };
    }
    case SELECT_USERSTORY: {
      return { ...state, currentUserStory: action.payload.Id };
    }

    default:
      return state;
  }
}

export function getHypothesisSelected(state) {
  return state.ui.currentHypothesis;
}
export function getUserStorySelected(state) {
  return state.ui.currentUserStory;
}
