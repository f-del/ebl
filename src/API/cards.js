import { cardStatus } from "../redux/modules/cards";

const mapping = card => {
  card = card._firestore !== undefined ? card.data() : card;
  return {
    ...card,
    Status:
      card.Status === "INPROGRESS" ? cardStatus.INPROGRESS : cardStatus.TODO
  };
};

function post(db) {
  return card => {
    return new Promise((resolve, reject) => {
      db.collection("cards")
        .add(card)
        .then(ref => {
          console.log("Document successfully written!");
          resolve({ id: ref.id });
        })
        .catch(error => {
          console.error("Error writing document: ", error);
          //dispatch(errorSave())
          reject(error);
        });
    });
  };
}

function get(db) {
  return type => {
    return new Promise((resolve, reject) => {
      db.collection("cards")
        .where("Type", "==", type)
        .get()
        .then(function(querySnapshot) {
          let cards = [];

          querySnapshot.forEach(function(doc) {
            let env = mapping(doc);

            cards.push(env);
          });

          resolve(cards);
        });
    });
  };
}

export default db => {
  return { Post: post(db), Get: get(db) };
};
