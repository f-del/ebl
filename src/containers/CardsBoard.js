import React from "react";
import { connect } from "react-redux";
import { retrieveAllCards, cardType } from "../redux/modules/cards";

const mapStateToProps = state => ({});

const mapDispatchToProps = (dispatch, ownsProp) => ({
  getCards: type => dispatch(retrieveAllCards(type || cardType.Task))
});

function CardsBoard({ type, getCards, children }) {
  getCards(type);
  return <div>{children}</div>;
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(CardsBoard);
