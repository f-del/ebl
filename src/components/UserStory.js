import React from "react";
import PropTypes from "prop-types";

import UiGridListTile from "@material-ui/core/GridListTile";
import CardsBoard from "../containers/CardsBoard";
import { cardType } from "../redux/modules/cards";
import CardListByIdList from "../containers/CardListByIdList";

function UserStory({ userStory }) {
  if (userStory === undefined) return null;

  return (
    <CardsBoard type={cardType.Task}>
      <UiGridListTile>
        {/* <CreateTaskCard userStoryId={userStory.Id} /> */}
      </UiGridListTile>
      <CardListByIdList list={userStory.Tasks} />
    </CardsBoard>
  );
}

UserStory.propTypes = {
  userStory: PropTypes.object
};

export default UserStory;
