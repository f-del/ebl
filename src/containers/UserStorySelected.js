import { connect } from "react-redux";
import { getUserStorySelected } from "../redux/modules/ui";
import { getCard } from "../redux/modules/cards";
import UserStory from "../components/UserStory";

const mapStateToProps = (state, ownsProp) => ({
  userStory: getCard(state, getUserStorySelected(state))
});

const mapDispatchToProps = (dispatch, ownsProp) => ({});

const UserStorySelected = connect(
  mapStateToProps,
  mapDispatchToProps
)(UserStory);

export default UserStorySelected;
