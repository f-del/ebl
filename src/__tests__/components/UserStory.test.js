import React from "react";
import { shallow, configure, mount } from "enzyme";
import Adapter from "enzyme-adapter-react-16";
import { createStore, applyMiddleware } from "redux";
import { Provider } from "react-redux";
import thunk from "redux-thunk";
import reducer from "../../redux/store/index";

import uiReducer from "../../redux/modules/ui";
import { entity_test_created } from "../datas";
import UserStorySelected from "../../containers/UserStorySelected";
import UserStory from "../../components/UserStory";
import { cardType } from "../../redux/modules/cards";
import CardsBoard from "../../containers/CardsBoard";

configure({ adapter: new Adapter() });
var store, api;

describe("Componant tests", () => {
  beforeEach(() => {
    jest.resetAllMocks();
    api = {
      Cards: {
        Post: jest.fn(),
        Get: jest.fn()
      }
    };

    store = createStore(
      reducer,
      applyMiddleware(thunk.withExtraArgument({ api }))
    );
  });

  test("Not already loaded UserStory", () => {
    const wrapper = shallow(<UserStory />);

    expect(wrapper.isEmptyRender()).toEqual(true);
  });

  test("UserStory setted externaly without attached Tasks", () => {
    const wrapper = shallow(
      <UserStory
        userStory={{ ...entity_test_created, Type: cardType.UserStory }}
      />
    );

    expect(wrapper.isEmptyRender()).toEqual(false);
    expect(
      wrapper
        .find(CardsBoard)
        .at(0)
        .props().type
    ).toStrictEqual(cardType.Task);
  });
});

describe("Containers tests", () => {
  beforeEach(() => {
    jest.resetAllMocks();
    store = createStore(
      reducer,
      applyMiddleware(thunk.withExtraArgument({ api }))
    );
  });
  test("Expect props onAction is defined", () => {
    uiReducer.getUserStorySelected = jest.fn().mockImplementationOnce(id => {
      return dispatch => {
        return entity_test_created;
      };
    });

    store = createStore(reducer, applyMiddleware(thunk));

    const wrapper = shallow(
      <Provider store={store}>
        <UserStorySelected />,
      </Provider>
    );

    expect(wrapper.props("userStory")).toBeDefined();
    expect(wrapper).toMatchSnapshot();
  });
});
