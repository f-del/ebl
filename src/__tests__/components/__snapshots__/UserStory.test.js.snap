// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Containers tests Expect props onAction is defined 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      },
      "storeState": Object {
        "cards": Object {
          "list": Array [],
          "status": Object {
            "HYPOTHESIS": Symbol(NULL),
            "TASK": Symbol(NULL),
            "USER-STORY": Symbol(NULL),
          },
        },
        "criterias": Object {
          "list": Object {
            "BASIC": Array [
              Object {
                "Id": "IdCritBas1",
                "Value": false,
              },
            ],
            "DEV": Array [
              Object {
                "Id": "IdCritDev1",
                "Text": "Development",
                "Value": false,
              },
              Object {
                "Id": "IdCritDev2",
                "Text": "Unit testing",
                "Value": false,
              },
            ],
          },
        },
        "personas": Object {
          "list": Object {
            "e6wTJIBsAw5puwwrNEsR": Object {
              "Details": "Meta persona, representing all collaborators working with Paul",
              "Name": "Feature Team",
              "Needs": Array [
                "To understand WHAT the Product Owner wants to be able to design collaborativly the HOW",
                "To manage easly the list of things to do to finish the most quickly the User Story",
              ],
              "Picture": "",
            },
            "uOCF2GLFiJl6fvxdlq4b": Object {
              "Details": "Product Owner for a new product, work in a startup",
              "Name": "Paul Deligny",
              "Needs": Array [
                "To have a clear understanding about the product vision & objectives",
                "To write real User Stories, ie have a Customer Driven thinking",
                "To share at the right time and with the rigth level of information the User Stories with his team",
                "To have a clear vision on the job he'll have to do AND the job already or currently done",
              ],
              "Picture": "",
            },
          },
        },
        "ui": Object {
          "currentHypothesis": undefined,
          "currentUserStory": undefined,
        },
      },
    }
  }
>
  <Connect(UserStory) />
  ,
</ContextProvider>
`;
